File: 4.8-bc/py/stage-3-hexagonal/app/billing/infra/http/handler.py
"""Billing HTTP handler — POST /subscriptions."""
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel

from app.billing.app.service import Service
from app.billing.domain.subscription import InvalidPlanError


class SubscribeReq(BaseModel):
    user_id: str
    plan: str


class SubscribeResp(BaseModel):
    subscription_id: str
    plan: str
    next_charge_at: datetime


def build_router(svc: Service) -> APIRouter:
    router = APIRouter(tags=["billing"])

    @router.post(
        "/subscriptions",
        status_code=status.HTTP_201_CREATED,
        response_model=SubscribeResp,
    )
    async def subscribe(req: SubscribeReq) -> SubscribeResp:
        if not req.plan:
            raise HTTPException(status_code=400, detail="invalid_request")
        try:
            uid = UUID(req.user_id)
        except (ValueError, TypeError) as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST, detail="invalid_user_id"
            ) from exc
        try:
            sub = await svc.subscribe(uid, req.plan)
        except InvalidPlanError as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={
                    "error": "billing.invalid_plan",
                    "message": "plan must be basic|pro|enterprise",
                },
            ) from exc
        return SubscribeResp(
            subscription_id=str(sub.id),
            plan=sub.plan,
            next_charge_at=sub.next_charge_at,
        )

    return router

File: 4.8-bc/go/stage-2-feature-first/app/billing/handler.go
package billing

import (
	"net/http"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-2-feature-first/shared/httputil"
)

type Handler struct {
	svc *Service
}

func NewHandler(svc *Service) *Handler {
	return &Handler{svc: svc}
}

func (h *Handler) Register(r chi.Router) {
	r.Post("/subscriptions", h.subscribe)
}

type subscribeReq struct {
	UserID string `json:"user_id"`
	Plan   string `json:"plan"`
}

func (h *Handler) subscribe(w http.ResponseWriter, r *http.Request) {
	var req subscribeReq
	if err := httputil.DecodeBody(r, &req); err != nil || req.Plan == "" {
		httputil.WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	uid, err := uuid.Parse(req.UserID)
	if err != nil {
		httputil.WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_user_id"})
		return
	}
	sub, err := h.svc.Subscribe(r.Context(), uid, req.Plan)
	if err != nil {
		httputil.WriteError(w, err)
		return
	}
	httputil.WriteJSON(w, http.StatusCreated, map[string]any{
		"subscription_id": sub.ID.String(),
		"plan":            sub.Plan,
		"next_charge_at":  sub.NextChargeAt,
	})
}
